src/passes/ConstantFieldPropagation.cpp

```diff
@@ -0,0 +1,393 @@
+    auto iter = infos->find(type.getHeapType());
+    if (iter != infos->end()) {
+      // There is information on this type, fetch it.
+      info = iter->second[curr->index];
+    }
+    if (!info.hasNoted()) {
+      // This field is never written at all. That means that we do not even
+      // construct any data of this type, and so it is a logic error to reach
+      // this location in the code. (Unless we are not in a closed-world
+      // situation, which we assume we are not in.) Replace this get with a
+      // trap. Note that we do not need to care about the nullability of the
+      // reference, as if it should have trapped, we are replacing it with
+      // another trap, which we allow to reorder (but we do need to care about
+      // side effects in the reference, so keep it around).
+      replaceCurrent(builder.makeSequence(builder.makeDrop(curr->ref),
+                                          builder.makeUnreachable()));
+      changedTypes = true;
+      return;
+    }
+
+    // If the value is not a constant, then it is unknown and we must give up.
+    if (!info.isConstant()) {
+      return;
+    }
+
+    // We can do this! Replace the get with a throw on a null reference (as the
+    // get would have done so), plus the constant value. (Leave it to further
+    // optimizations to get rid of the ref.)
+    replaceCurrent(builder.makeSequence(
+      builder.makeDrop(builder.makeRefAs(RefAsNonNull, curr->ref)),
+      builder.makeConstantExpression(info.getConstantValue())));
+    changedTypes = true;
+  }
+
+  void doWalkFunction(Function* func) {
+    WalkerPass<PostWalker<FunctionOptimizer>>::doWalkFunction(func);
+
+    // If we changed any types, we need to update parents.
+    if (changedTypes) {
+      ReFinalize().walkFunctionInModule(func, getModule());
+    }
+  }
+
+private:
+  StructValuesMap* infos;
+
+  bool changedTypes = false;
+};
+
+struct ConstantFieldPropagation : public Pass {
+  void run(PassRunner* runner, Module* module) override {
+    if (getTypeSystem() != TypeSystem::Nominal) {
+      Fatal() << "ConstantFieldPropagation requires nominal typing";
+    }
+
+    // Find and analyze all writes inside each function.
+    FunctionStructValuesMap functionInfos;
+    for (auto& func : module->functions) {
+      // Initialize the data for each function, so that we can operate on this
+      // structure in parallel without modifying it.
+      functionInfos[func.get()];
+    }
+    Scanner scanner(&functionInfos);
+    scanner.run(runner, module);
+    scanner.walkModuleCode(module);
+
+    // Combine the data from the functions.
+    StructValuesMap combinedInfos;
+    for (auto& kv : functionInfos) {
+      StructValuesMap& infos = kv.second;
+      for (auto& kv : infos) {
+        auto type = kv.first;
+        auto& info = kv.second;
+        auto& combinedInfo = combinedInfos[type];
+        for (Index i = 0; i < info.size(); i++) {
+          combinedInfo[i].combine(info[i]);
```

Comment:
nit: the `combinedInfo` intermediate variable doesn't pull its weight IMO.
```suggestion
          combinedInfos[type][i].combine(info[i]);
```